zfs/zfs.go

```diff
@@ -101,37 +107,77 @@ func (z *Zfs) Recv(fs string, sendCommand *exec.Cmd) error {
 	return nil
 }
 
-func (z *Zfs) SendIncremental(fs string, prev, snap string) *exec.Cmd {
-	snapfqn := fmt.Sprintf("%s@%s", fs, snap)
-	return z.exec("/sbin/zfs", "send", "-i", prev, snapfqn)
+// Send performs the `zfs send` command
+func (z *Zfs) Send(fs string, snap string) *exec.Cmd {
+	return z.exec("/sbin/zfs", "send", fmt.Sprintf("%s@%s", fs, snap))
 }
 
-func DoSync(_from, _to Fs) error {
-	from, _ := _from.(*fs) // Ugly, to remove
-	to, _ := _to.(*fs)     // Ugly, to remove
+// SendIncremental performs the `zfs send -i` command
+func (z *Zfs) SendIncremental(fs string, previous, current string) *exec.Cmd {
+	return z.exec("/sbin/zfs", "send", "-i",
+		fmt.Sprintf("@%s", previous),
+		fmt.Sprintf("%s@%s", fs, current),
+	)
+}
 
-	lastLocal := to.snaps[len(to.snaps)-1]
+// Returns the index of the last common snapshot
+func lastCommonSnapshotIndex(listA, listB []string) int {
+	result := -1
 
-	remoteIndex := indexOf(from.snaps, lastLocal)
+	for i, name := range listA {
+		if indexOf(listB, name) != -1 {
+			result = i
+		}
+	}
 
-	missing := from.snaps[remoteIndex+1:]
+	return result
+}
 
-	if len(missing) == 0 {
-		fmt.Println("Nothing to do")
-		return nil
-	}
+// DoSync create missing file systems on the destination and transfers missing snapshots
+func DoSync(from, to *Fs, recursive bool) error {
```

Comment:
Recieving data will always overwrite the destination filesystem changes.
Two possibilities:
- When there are snapshot names in common, the only data loss would be the changes made locally after the last snapshot sync done. I suppose we can expect the user to be aware that the modifications made to the destination filesystem of the synchronisation without doing a clone will not be kept after the next sync.
As the snapshots sent here are incremental, we can't completely overwrite the destination by mistake as the recv command will fail if the common snapshot does not contain the same data
- When there are no snapshot in common, the data is not sent incrementally so we ALWAYS overwrite the destination filesystem WITHOUT warning the user about it. This can lead to dramatic data losses and we definitely don't want that.

So either:
- Add a `-F` flag to the `sync` command so the user is aware that data can be lost
- Ask for user permission to execute L161: `to.Recv(from.Send(from.snaps[0]))` by asking confirmation (like ssh when connecting to a new server)